File: HikvisionService/Controllers/StoreController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using HikvisionService.Data;
using HikvisionService.Models;
using System.ComponentModel.DataAnnotations;

namespace HikvisionService.Controllers;

public class StoreController : Controller
{
    private readonly HikvisionDbContext _context;
    private readonly ILogger<StoreController> _logger;

    public StoreController(HikvisionDbContext context, ILogger<StoreController> logger)
    {
        _context = context;
        _logger = logger;
    }

    // GET: /Store
    public async Task<IActionResult> Index()
    {
        try
        {
            var stores = await _context.Stores
                .Include(s => s.Cameras)
                .OrderBy(s => s.Name)
                .ToListAsync();

            return View(stores);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting stores");
            TempData["ErrorMessage"] = "An error occurred while retrieving stores.";
            return View(new List<Store>());
        }
    }

    // GET: /Store/Details/5
    public async Task<IActionResult> Details(long id)
    {
        try
        {
            var store = await _context.Stores
                .Include(s => s.Cameras)
                .FirstOrDefaultAsync(s => s.Id == id);

            if (store == null)
            {
                TempData["ErrorMessage"] = $"Store with ID {id} not found.";
                return RedirectToAction(nameof(Index));
            }

            return View(store);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting store {StoreId}", id);
            TempData["ErrorMessage"] = "An error occurred while retrieving the store.";
            return RedirectToAction(nameof(Index));
        }
    }

    // GET: /Store/Create
    public IActionResult Create()
    {
        return View();
    }

    // POST: /Store/Create
    [HttpPost]
    [ValidateAntiForgeryToken]
    public async Task<IActionResult> Create(StoreModel model)
    {
        if (ModelState.IsValid)
        {
            try
            {
                // Check if store name already exists
                if (await _context.Stores.AnyAsync(s => s.Name == model.Name))
                {
                    ModelState.AddModelError("Name", "Store name already exists");
                    return View(model);
                }

                var store = new Store
                {
                    Name = model.Name,
                    CreatedAt = DateTime.UtcNow,
                    UpdatedAt = DateTime.UtcNow
                };

                _context.Stores.Add(store);
                await _context.SaveChangesAsync();

                _logger.LogInformation("Created store: {StoreName} with ID: {StoreId}", store.Name, store.Id);
                TempData["SuccessMessage"] = "Store created successfully.";

                return RedirectToAction(nameof(Index));
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error creating store");
                ModelState.AddModelError("", "An error occurred while creating the store.");
            }
        }

        return View(model);
    }

    // GET: /Store/Edit/5
    public async Task<IActionResult> Edit(long id)
    {
        try
        {
            var store = await _context.Stores.FindAsync(id);
            if (store == null)
            {
                TempData["ErrorMessage"] = $"Store with ID {id} not found.";
                return RedirectToAction(nameof(Index));
            }

            var viewModel = new StoreModel
            {
                Id = store.Id,
                Name = store.Name
            };

            return View(viewModel);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting store {StoreId} for edit", id);
            TempData["ErrorMessage"] = "An error occurred while retrieving the store.";
            return RedirectToAction(nameof(Index));
        }
    }

    // POST: /Store/Edit/5
    [HttpPost]
    [ValidateAntiForgeryToken]
    public async Task<IActionResult> Edit(long id, StoreModel model)
    {
        if (id != model.Id)
        {
            TempData["ErrorMessage"] = "Invalid store ID.";
            return RedirectToAction(nameof(Index));
        }

        if (ModelState.IsValid)
        {
            try
            {
                var store = await _context.Stores.FindAsync(id);
                if (store == null)
                {
                    TempData["ErrorMessage"] = $"Store with ID {id} not found.";
                    return RedirectToAction(nameof(Index));
                }

                // Check if new name already exists (excluding current store)
                if (await _context.Stores.AnyAsync(s => s.Name == model.Name && s.Id != id))
                {
                    ModelState.AddModelError("Name", "Store name already exists");
                    return View(model);
                }

                store.Name = model.Name;
                store.UpdatedAt = DateTime.UtcNow;

                await _context.SaveChangesAsync();

                _logger.LogInformation("Updated store {StoreId}: {StoreName}", store.Id, store.Name);
                TempData["SuccessMessage"] = "Store updated successfully.";

                return RedirectToAction(nameof(Index));
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error updating store {StoreId}", id);
                ModelState.AddModelError("", "An error occurred while updating the store.");
            }
        }

        return View(model);
    }

    // GET: /Store/Delete/5
    public async Task<IActionResult> Delete(long id)
    {
        try
        {
            var store = await _context.Stores
                .Include(s => s.Cameras)
                .FirstOrDefaultAsync(s => s.Id == id);

            if (store == null)
            {
                TempData["ErrorMessage"] = $"Store with ID {id} not found.";
                return RedirectToAction(nameof(Index));
            }

            return View(store);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting store {StoreId} for delete", id);
            TempData["ErrorMessage"] = "An error occurred while retrieving the store.";
            return RedirectToAction(nameof(Index));
        }
    }

    // POST: /Store/Delete/5
    [HttpPost, ActionName("Delete")]
    [ValidateAntiForgeryToken]
    public async Task<IActionResult> DeleteConfirmed(long id)
    {
        try
        {
            var store = await _context.Stores
                .Include(s => s.Cameras)
                .FirstOrDefaultAsync(s => s.Id == id);

            if (store == null)
            {
                TempData["ErrorMessage"] = $"Store with ID {id} not found.";
                return RedirectToAction(nameof(Index));
            }

            // Check if store has cameras
            if (store.Cameras.Any())
            {
                TempData["ErrorMessage"] = "Cannot delete store with associated cameras. Please delete cameras first.";
                return RedirectToAction(nameof(Delete), new { id });
            }

            _context.Stores.Remove(store);
            await _context.SaveChangesAsync();

            _logger.LogInformation("Deleted store {StoreId}: {StoreName}", store.Id, store.Name);
            TempData["SuccessMessage"] = "Store deleted successfully.";

            return RedirectToAction(nameof(Index));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting store {StoreId}", id);
            TempData["ErrorMessage"] = "An error occurred while deleting the store.";
            return RedirectToAction(nameof(Index));
        }
    }

    // View Models
    public class StoreModel
    {
        public long Id { get; set; }

        [Required(ErrorMessage = "Store name is required")]
        [StringLength(100, ErrorMessage = "Store name cannot be longer than 100 characters")]
        public string Name { get; set; } = string.Empty;
    }
}

File: laravel/app/Http/Controllers/StoreController.php
<?php

namespace App\Http\Controllers;

use App\Models\Store;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;

class StoreController extends Controller
{
    public function index()
    {
        try {
            $stores = Store::withCount('cameras')->orderBy('name')->get();
            return view('stores.index', compact('stores'));
        } catch (\Exception $ex) {
            Log::error('Error getting stores', ['error' => $ex->getMessage()]);
            return redirect()->back()->with('error', 'An error occurred while retrieving stores.');
        }
    }

    public function show($id)
    {
        try {
            $store = Store::with('cameras')->findOrFail($id);
            return view('stores.show', compact('store'));
        } catch (\Exception $ex) {
            Log::error("Error getting store {$id}", ['error' => $ex->getMessage()]);
            return redirect()->route('stores.index')->with('error', 'Store not found.');
        }
    }

    public function create()
    {
        return view('stores.create');
    }

    public function store(Request $request)
    {
        $request->validate([
            'name' => 'required|string|max:255',
            'address' => 'nullable|string|max:500',
            'phone' => 'nullable|string|max:20',
            'email' => 'nullable|email|max:100',
        ]);

        try {
            Store::create([
                'name' => $request->name,
                'address' => $request->address,
                'phone' => $request->phone,
                'email' => $request->email,
            ]);

            Log::info('Created store', ['name' => $request->name]);
            return redirect()->route('stores.index')->with('success', 'Store created successfully.');
        } catch (\Exception $ex) {
            Log::error('Error creating store', ['error' => $ex->getMessage()]);
            return back()->withInput()->with('error', 'An error occurred while creating the store.');
        }
    }

    public function edit($id)
    {
        try {
            $store = Store::findOrFail($id);
            return view('stores.edit', compact('store'));
        } catch (\Exception $ex) {
            Log::error("Error getting store {$id} for edit", ['error' => $ex->getMessage()]);
            return redirect()->route('stores.index')->with('error', 'Store not found.');
        }
    }

    public function update(Request $request, $id)
    {
        $request->validate([
            'name' => 'required|string|max:255',
            'address' => 'nullable|string|max:500',
            'phone' => 'nullable|string|max:20',
            'email' => 'nullable|email|max:100',
        ]);

        try {
            $store = Store::findOrFail($id);

            $store->update([
                'name' => $request->name,
                'address' => $request->address,
                'phone' => $request->phone,
                'email' => $request->email,
            ]);

            Log::info('Updated store', ['id' => $id, 'name' => $request->name]);
            return redirect()->route('stores.index')->with('success', 'Store updated successfully.');
        } catch (\Exception $ex) {
            Log::error("Error updating store {$id}", ['error' => $ex->getMessage()]);
            return back()->withInput()->with('error', 'An error occurred while updating the store.');
        }
    }

    public function destroy($id)
    {
        try {
            $store = Store::with('cameras')->findOrFail($id);

            if ($store->cameras->count() > 0) {
                return redirect()->route('stores.index')->with('error', 'Cannot delete store with associated cameras. Please delete or reassign the cameras first.');
            }

            $storeName = $store->name;
            $store->delete();

            Log::info('Deleted store', ['id' => $id, 'name' => $storeName]);
            return redirect()->route('stores.index')->with('success', 'Store deleted successfully.');
        } catch (\Exception $ex) {
            Log::error("Error deleting store {$id}", ['error' => $ex->getMessage()]);
            return redirect()->route('stores.index')->with('error', 'An error occurred while deleting the store.');
        }
    }
}
